pytest.skip inside doctests does not work (UnexpectedException)

[`pytest.skip` docstring](https://docs.pytest.org/en/latest/reference.html#pytest-skip) says "This function should be called only during testing (setup, call or teardown)".  However, calling `pytest.skip` inside a doctest is counted as a failure:

```console
$ ls
env  test.txt

$ cat test.txt
>>> import pytest
>>> pytest.skip("Skip doctest")

$ env/bin/pytest
================== test session starts ==================
platform linux -- Python 3.7.0, pytest-4.3.0, py-1.8.0, pluggy-0.9.0
rootdir: /tmp/tmp.YVVwv7GVVo, inifile:
collected 1 item

test.txt F                                        [100%]

======================= FAILURES ========================
__________________ [doctest] test.txt ___________________
001 >>> import pytest
002 >>> pytest.skip("Skip doctest")
UNEXPECTED EXCEPTION: Skip doctest
Traceback (most recent call last):

  File "/usr/lib64/python3.7/doctest.py", line 1329, in __run
    compileflags, 1), test.globs)

  File "<doctest test.txt[1]>", line 1, in <module>

  File "/tmp/tmp.YVVwv7GVVo/env/lib/python3.7/site-packages/_pytest/outcomes.py", line 98, in skip
    raise Skipped(msg=msg, allow_module_level=allow_module_level)

Skipped: Skip doctest

/tmp/tmp.YVVwv7GVVo/test.txt:2: UnexpectedException
=============== 1 failed in 0.01 seconds ================
```

Would it be possible to make `pytest.skip` work inside doctests?


<details>
 <summary>System information:</summary>

```console
$ env/bin/pip list
Package        Version
-------------- -------
atomicwrites   1.3.0
attrs          19.1.0
more-itertools 6.0.0
pip            19.0.3
pluggy         0.9.0
py             1.8.0
pytest         4.3.0
setuptools     40.8.0
six            1.12.0
wheel          0.33.1

$ env/bin/pytest --version
This is pytest version 4.3.0, imported from /tmp/tmp.YVVwv7GVVo/env/lib/python3.7/site-packages/pytest.py

$ uname -a
Linux tkf-xps 4.14.61-1-lts #1 SMP Mon Aug 6 21:07:39 CEST 2018 x86_64 GNU/Linux

$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
ID_LIKE=archlinux
ANSI_COLOR="0;36"
HOME_URL="https://www.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
```

</details>

--- a/AUTHORS
+++ b/AUTHORS
@@ -222,6 +222,7 @@ Steffen Allner
 Stephan Obermann
 Sven-Hendrik Haase
 Tadek Teleżyński
+Takafumi Arakaki
 Tarcisio Fischer
 Tareq Alayan
 Ted Xiao

--- a/changelog/4911.feature.rst
+++ b/changelog/4911.feature.rst
@@ -1,0 +1,1 @@
+Doctest can be now skipped dynamically with `pytest.skip`.

--- a/changelog/4911.feature.rst
+++ b/changelog/4911.feature.rst
@@ -1,1 +1,1 @@
-Doctest can be now skipped dynamically with `pytest.skip`.
+Doctests can be skipped now dynamically using ``pytest.skip()``.

--- a/src/_pytest/doctest.py
+++ b/src/_pytest/doctest.py
@@ -15,6 +15,7 @@ from _pytest._code.code import ReprFileLocation
 from _pytest._code.code import TerminalRepr
 from _pytest.compat import safe_getattr
 from _pytest.fixtures import FixtureRequest
+from _pytest.outcomes import Skipped
 
 DOCTEST_REPORT_CHOICE_NONE = "none"
 DOCTEST_REPORT_CHOICE_CDIFF = "cdiff"
@@ -153,6 +154,8 @@ def _init_runner_class():
                 raise failure
 
         def report_unexpected_exception(self, out, test, example, exc_info):
+            if isinstance(exc_info[1], Skipped):
+                raise exc_info[1]
             failure = doctest.UnexpectedException(test, example, exc_info)
             if self.continue_on_failure:
                 out.append(failure)

--- a/src/_pytest/outcomes.py
+++ b/src/_pytest/outcomes.py
@@ -81,7 +81,7 @@ def skip(msg="", **kwargs):
 
     This function should be called only during testing (setup, call or teardown) or
     during collection by using the ``allow_module_level`` flag.  This function can
-    be called in doctest as well.
+    be called in doctests as well.
 
     :kwarg bool allow_module_level: allows this function to be called at
         module level, skipping the rest of the module. Default to False.

--- a/src/_pytest/outcomes.py
+++ b/src/_pytest/outcomes.py
@@ -80,7 +80,8 @@ def skip(msg="", **kwargs):
     Skip an executing test with the given message.
 
     This function should be called only during testing (setup, call or teardown) or
-    during collection by using the ``allow_module_level`` flag.
+    during collection by using the ``allow_module_level`` flag.  This function can
+    be called in doctest as well.
 
     :kwarg bool allow_module_level: allows this function to be called at
         module level, skipping the rest of the module. Default to False.
@@ -89,6 +90,9 @@ def skip(msg="", **kwargs):
         It is better to use the :ref:`pytest.mark.skipif ref` marker when possible to declare a test to be
         skipped under certain conditions like mismatching platforms or
         dependencies.
+        Similarly, use ``# doctest: +SKIP`` directive (see `doctest.SKIP
+        <https://docs.python.org/3/library/doctest.html#doctest.SKIP>`_)
+        to skip a doctest statically.
     """
     __tracebackhide__ = True
     allow_module_level = kwargs.pop("allow_module_level", False)

--- a/src/_pytest/outcomes.py
+++ b/src/_pytest/outcomes.py
@@ -90,7 +90,7 @@ def skip(msg="", **kwargs):
         It is better to use the :ref:`pytest.mark.skipif ref` marker when possible to declare a test to be
         skipped under certain conditions like mismatching platforms or
         dependencies.
-        Similarly, use ``# doctest: +SKIP`` directive (see `doctest.SKIP
+        Similarly, use the ``# doctest: +SKIP`` directive (see `doctest.SKIP
         <https://docs.python.org/3/library/doctest.html#doctest.SKIP>`_)
         to skip a doctest statically.
     """

--- a/testing/test_doctest.py
+++ b/testing/test_doctest.py
@@ -188,6 +188,18 @@ class TestDoctests(object):
             ]
         )
 
+    def test_doctest_skip(self, testdir):
+        testdir.maketxtfile(
+            """
+            >>> 1
+            1
+            >>> import pytest
+            >>> pytest.skip("")
+        """
+        )
+        result = testdir.runpytest("--doctest-modules")
+        result.stdout.fnmatch_lines(["*1 skipped*"])
+
     def test_docstring_partial_context_around_error(self, testdir):
         """Test that we show some context before the actual line of a failing
         doctest.

